components/logins/src/encryption.rs

```diff
@@ -26,24 +26,140 @@
+}
+
+impl ManagedEncryptorDecryptor {
+    pub fn new(key_manager: Arc<dyn KeyManager>) -> Self {
+        Self { key_manager }
+    }
+}
 
-pub type EncryptorDecryptor = jwcrypto::EncryptorDecryptor<Error>;
+impl EncryptorDecryptor for ManagedEncryptorDecryptor {
+    fn encrypt(&self, clearbytes: Vec<u8>) -> ApiResult<Vec<u8>> {
+        let keybytes = self
+            .key_manager
+            .get_key()
+            .map_err(|_| LoginsApiError::MissingKey)?;
+        let key = std::str::from_utf8(&keybytes).map_err(|_| LoginsApiError::InvalidKey)?;
+
+        let encdec = jwcrypto::EncryptorDecryptor::new(key)
+            .map_err(|_: jwcrypto::EncryptorDecryptorError| LoginsApiError::InvalidKey)?;
+
+        let cleartext =
+            std::str::from_utf8(&clearbytes).map_err(|e| LoginsApiError::EncryptionFailed {
+                reason: e.to_string(),
+            })?;
+        encdec
+            .encrypt(cleartext, "encrypt SecureLoginFields")
+            .map_err(
+                |e: jwcrypto::EncryptorDecryptorError| LoginsApiError::EncryptionFailed {
+                    reason: e.to_string(),
+                },
+            )
+            .map(|text| text.into())
+    }
+
+    fn decrypt(&self, cipherbytes: Vec<u8>) -> ApiResult<Vec<u8>> {
+        let keybytes = self
+            .key_manager
+            .get_key()
+            .map_err(|_| LoginsApiError::MissingKey)?;
+        let key = std::str::from_utf8(&keybytes).map_err(|_| LoginsApiError::InvalidKey)?;
+
+        let encdec = jwcrypto::EncryptorDecryptor::new(key)
+            .map_err(|_: jwcrypto::EncryptorDecryptorError| LoginsApiError::InvalidKey)?;
+
+        let ciphertext =
+            std::str::from_utf8(&cipherbytes).map_err(|e| LoginsApiError::DecryptionFailed {
+                reason: e.to_string(),
+            })?;
+        encdec
+            .decrypt(ciphertext, "decrypt SecureLoginFields")
+            .map_err(
+                |e: jwcrypto::EncryptorDecryptorError| LoginsApiError::DecryptionFailed {
+                    reason: e.to_string(),
+                },
+            )
+            .map(|text| text.into())
+    }
+}
+
+/// Consumers can implement the KeyManager in combination with the ManagedEncryptorDecryptor to hand
+/// over the encryption key whenever encryption or decryption happens.
+pub trait KeyManager: Send + Sync {
+    fn get_key(&self) -> ApiResult<Vec<u8>>;
+}
+
+/// Last but not least we provide a StaticKeyManager, which can be
+/// used in cases where there is a single key during runtime, for example in tests.
+pub struct StaticKeyManager {
+    key: String,
+}
+
+impl StaticKeyManager {
+    pub fn new(key: String) -> Self {
+        Self { key }
+    }
+}
+
+impl KeyManager for StaticKeyManager {
+    #[handle_error(Error)]
+    fn get_key(&self) -> ApiResult<Vec<u8>> {
+        Ok(self.key.as_bytes().into())
+    }
+}
 
 #[handle_error(Error)]
 pub fn create_canary(text: &str, key: &str) -> ApiResult<String> {
```

Comment:
The canary story can probably be better too - the intent there is to better work with keys which are our of our control. IIUC there's no equivalent thing on desktop. Maybe that concept could be optionally added to the traits? This might be good fodder for a followup.

Reply:
Yes, very good idea!

Reply:
We would have to change our udl so that it is automatically generated with [procmacros](https://mozilla.github.io/uniffi-rs/0.27/proc_macro/index.html), which I would very much like, right?

Reply:
I don't quite understand the question, but as a meta comment, I think we are in the position to use procmacros (see #6413), so if you'd prefer killing the UDL entirely I think it would work fine.